### PR TITLE
Simplify windows builtin functions return type

### PR DESCRIPTION
**Pull request by comphead**
## Which issue does this PR close?

Closes #.

## Rationale for this change
Before the PR the Datafusion derives the output datatypes for builtin functions twice. 

Initially set in `BuiltInWindowFunction::return_type` the output datatype redefined in other places:
- in built in function implementations, like row_number, rank, etc
- on physical planner phase like lag, lead

This PR is to unify output types to be defined in 1 single place and be transferred through the schema. 

Another good benefit of it is other systems that embed DataFusion as an engine can now use its datatypes by constructing expected schema.

## What changes are included in this PR?
Datatype definition unified, covered with new tests, some other minor optimizations

## Are these changes tested?
Yes

## Are there any user-facing changes?
No

### CODE REVIEW

File: datafusion/physical-expr/src/window/cume_dist.rs [hunk 1]
```diff
@@ -34,11 +34,15 @@ use std::sync::Arc;
 #[derive(Debug)]
 pub struct CumeDist {
     name: String,
+    data_type: DataType,
```
Comment (alamb) [inline datafusion/physical-expr/src/window/cume_dist.rs:37]: I think it would help to note that this is the output type, not the argument type

```suggestion
    /// Output data type
    data_type: DataType,
```

File: datafusion/physical-expr/src/window/rank.rs [hunk 2]
```diff
@@ -89,11 +93,7 @@ impl BuiltInWindowFunctionExpr for Rank {
 
     fn field(&self) -> Result<Field> {
         let nullable = false;
-        let data_type = match self.rank_type {
-            RankType::Basic | RankType::Dense => DataType::UInt64,
-            RankType::Percent => DataType::Float64,
-        };
-        Ok(Field::new(self.name(), data_type, nullable))
+        Ok(Field::new(self.name(), self.data_type.clone(), nullable))
```
Comment (alamb) [inline datafusion/physical-expr/src/window/rank.rs:96]: ❤️

File: datafusion/sqllogictest/test_files/window.slt [hunk 3]
```diff
@@ -3906,3 +3906,69 @@ ProjectionExec: expr=[sn@0 as sn, ts@1 as ts, currency@2 as currency, amount@3 a
 --BoundedWindowAggExec: wdw=[SUM(table_with_pk.amount) ORDER BY [table_with_pk.sn ASC NULLS LAST] ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW: Ok(Field { name: "SUM(table_with_pk.amount) ORDER BY [table_with_pk.sn ASC NULLS LAST] ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW", data_type: Float64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Rows, start_bound: Preceding(UInt64(NULL)), end_bound: CurrentRow }], mode=[Sorted]
 ----SortExec: expr=[sn@0 ASC NULLS LAST]
 ------MemoryExec: partitions=1, partition_sizes=[1]
+
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/window.slt:3909]: These tests all seem to pass for me on main as well (without the changes in this PR). Is that expected?

File: datafusion/core/src/physical_planner.rs [hunk 4]
```diff
@@ -719,14 +720,16 @@ impl DefaultPhysicalPlanner {
                     }
 
                     let logical_input_schema = input.schema();
-                    let physical_input_schema = input_exec.schema();
+                    // Extend the schema to include window expression fields as builtin window functions derives its datatype from incoming schema
```
Comment (alamb) [inline datafusion/core/src/physical_planner.rs:723]: Shouldn't `input.schema()` reflect all the columns that the input produces?

Or does the `WindowAggExec` create new columns "internally" by evaluating the window expressions?

**Review by alamb — APPROVED** [review]
> This PR is to unify output types to be defined in 1 single place and be transferred through the schema.

Thank you @comphead  -- this change looks good to me. 

Before we merge it, however, I think we should give @mustafasrepo / @ozankabak  a chance to comment as I think they are familiar with this code

**Comment by ozankabak** [comment]
@mustafasrepo PTAL

File: datafusion/core/src/physical_planner.rs (hunk 4 above)
Comment (comphead) [reply datafusion/core/src/physical_planner.rs:723]: input schema is schema from previous plan node(no window expressions). afaik windows expression column being added separately

https://github.com/apache/arrow-datafusion/blob/0116e2a9b4a3ed4491802e19195769b96b7a971a/datafusion/expr/src/logical_plan/plan.rs#L2045

File: datafusion/sqllogictest/test_files/window.slt (hunk 3 above)
Comment (comphead) [reply datafusion/sqllogictest/test_files/window.slt:3909]: Yes, tests checked the functionality is not regressed.

File: datafusion/physical-expr/src/window/cume_dist.rs (hunk 1 above)
Comment (comphead) [reply datafusion/physical-expr/src/window/cume_dist.rs:37]: will do for all functions

**Comment by mustafasrepo** [comment]
I think this PR add two functionality. Which are
- Removing hard coded output types from builtin window functions so that all handled using `return_type` API.
- Removing `physical_input_schema` from `create_window_expr` function.
I think first change is better. However, second change is misleading. With the changes in this PR, given `logical_input_schema` is no longer actual input schema, but really window schema. 
However, I think we can merge this PR as is, Then I will file a PR to retract 2nd part. We can discuss pros and cons of it in that PR.

**Comment by mustafasrepo** [comment]
Filed the [PR8945](https://github.com/apache/arrow-datafusion/pull/8945) for retracting second change

File: datafusion/core/src/physical_planner.rs [hunk 5]
```diff
@@ -719,14 +720,16 @@ impl DefaultPhysicalPlanner {
...
+                    window_fields.extend_from_slice(&exprlist_to_fields(window_expr.iter(), input)?);
+                    let extended_schema = &DFSchema::new_with_metadata(window_fields, HashMap::new())?;
                     let window_expr = window_expr
                         .iter()
                         .map(|e| {
                             create_window_expr(
                                 e,
-                                logical_input_schema,
-                                &physical_input_schema,
+                                extended_schema,
```
Comment (viirya) [inline datafusion/core/src/physical_planner.rs:732]: Why do you need to recompute this `extended_schema` which looks like the same of window operator's schema.

File: datafusion/core/src/physical_planner.rs [hunk 6]
```diff
@@ -1598,21 +1602,14 @@ pub fn create_window_expr_with_name(
 pub fn create_window_expr(
     e: &Expr,
     logical_input_schema: &DFSchema,
```
Comment (viirya) [inline datafusion/core/src/physical_planner.rs:1604]: This is misleading, it is actually window's schema, not input schema now.

File: datafusion/core/src/physical_planner.rs [hunk 7]
```diff
@@ -1572,7 +1575,8 @@ pub fn create_window_expr_with_name(
                     create_physical_sort_expr(e, logical_input_schema, execution_props)
```
Comment (viirya) [inline datafusion/core/src/physical_planner.rs:1575]: Hmm, this looks incorrect as you use window's schema as input schema. Although window's schema is input schema + window functions output, it is why this change still makes thing work. But it is actually misleading for readers and probably cause of potential bugs.

File: datafusion/core/src/physical_planner.rs (hunk 5 above)
Comment (comphead) [reply datafusion/core/src/physical_planner.rs:732]: Filed https://github.com/apache/arrow-datafusion/pull/8955
